### ISSUE TITLE
Wrong automatic gradient for Linear-Layer (Autograd/gradcheck)

### ISSUE THREAD

**Issue by Twittadrock** [pytorch/pytorch#5351]
Minimal Example:
```
import torch.utils.data
import torch.autograd
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.autograd import Variable
from torch.autograd import gradcheck

torch.manual_seed(123)
input = Variable(torch.rand(1,2), requires_grad = True)
test = gradcheck(nn.Linear(2,1), (input,), eps=1e-6, atol=1e-3)
print(test)
```
Output:
```
 numerical:(
-0.3576
 0.2682
[torch.FloatTensor of size 2x1]
,)
analytical:(
-0.3512
 0.2667
[torch.FloatTensor of size 2x1]
,)

```
My Environment:
```
Python: (3, 5, 4)
CuDNN: 7005
CUDA: 8.0.61
Torch: 0.3.1.post2
Numpy: 1.13.3


```

**Comment by matherm** [comment]
Are we overseeing something essential or is there really a bug in the current release? Sorry for the brevity, I forwarded the issue from a colleague via mobile.

Beste wishes, 
Matthias

**Comment by apaszke** [comment]
You're running grad checks with single precision. If you add `.double()` calls it will pass.

**Comment by Naman-ntc** [comment]
@apaszke Is there any way that gradcheck passes for float operations. I am working on adversarial attacks and due to this precision loss in adversarial attacks over iterations I am getting different output across iterations. If I use DoubleTensors it works but becomes very slow (3-4 times slower)
